src/api/app/views/webui/request/_sourcediff_tab.html.haml

```diff
@@ -7,43 +7,50 @@
-                                                                                        filename: filename,
-                                                                                        file: files[filename],
-                                                                                        index: file_index,
-                                                                                        last: sourcediff['filenames'].last == filename,
-                                                                                        package: action[:spkg] }
+          - if sourcediff[:error]
+            %p
+              %i.error
+                = sourcediff[:error]
           - else
-            .mb-2
-              %p.lead
-                No source changes.
-      .col-lg-4
-        = render partial: 'webui/shared/buildresult_box',
-                 locals: { project: action[:sprj], package: action[:spkg], bs_request: bs_request, index: index }
-        - if sourcediff['issues'].present?
-          = render partial: 'issues_table', locals: { issues: sourcediff['issues'] }
+            - if action[:sourcediff].length > 1
+              %h4
+                #{diff_label(sourcediff['new'])} – #{diff_label(sourcediff['old'])}
+            - files = sourcediff['files']
+            - diff_params = diff_data(action[:type], sourcediff)
+            - if sourcediff['filenames'].present?
+              - sourcediff['filenames'].each_with_index do |filename, file_index|
+                - diff_params[:filename] = filename
+                -# Deleted files can not be linked. We have to hide them.
+                - file_view_path = package_view_file_path(diff_params) unless files[filename]['state'] == 'deleted'
+                .mb-2
+                  = render partial: 'webui/package/revision_diff_detail', locals: { file_view_path: file_view_path,
+                                                                                          filename: filename,
+                                                                                          file: files[filename],
+                                                                                          index: file_index,
+                                                                                          last: sourcediff['filenames'].last == filename,
+                                                                                          package: action[:spkg] }
+            - else
+              .mb-2
+                %p.lead
+                  No source changes.
+        .col-lg-4
+          = render partial: 'webui/shared/buildresult_box',
+                  locals: { project: action[:sprj], package: action[:spkg], bs_request: bs_request, index: index }
+          - if sourcediff['issues'].present?
+            = render partial: 'issues_table', locals: { issues: sourcediff['issues'] }
```

Comment:
You can move a lot of this logic into the `RequestActionController` and maybe view helpers / decorators so the view is as simple as possible

Reply:
Please keep in mind that in the not so far future we are going for a re-design of this page which will probably include a lot of refactoring. So doing large refactorings now might be not the right time.

Reply:
Please keep in mind that in the not so far future we are going for a re-design of this page which will probably include a lot of refactoring. So doing large refactorings now might be not the right time.

Reply:
Hm, good point. Light refactoring is fine for me, just removing "the noise": those `files`, `diff_params` variables and some of those `if`'s because logic in haml is not fun...

Reply:
We'll come back to this later when we start to refactor this part.